### ISSUE TITLE
Proposal: RadialController injection input

### ISSUE THREAD

**Issue by hansmbakker** [microsoft/ProjectReunion#54]
# Proposal: RadialController injection input

## Summary
There is a [preview API to inject HID input](https://docs.microsoft.com/en-us/uwp/api/windows.ui.input.preview.injection?view=winrt-18362). Currently the input types are restricted to
- keyboard
- mouse
- gamepad
- pen

It would be great if RadialController (Surface Dial) input could be injected, too. This would enable making non-compatible devices act as RadialController devices via a user-space app instead of writing a driver.

## Rationale

* I put the request in this repository since the [current preview API](https://docs.microsoft.com/en-us/uwp/api/windows.ui.input.preview.injection?view=winrt-18362) is in the `Windows.UI.*` namespace, and there is no good alternative where feedback feels heard
* Would enable people to write e.g. emulators for RadialController without having to resort to driver development (which is a barrier)

## Open Questions

If there is a better place to request this, please let me know. I just do not feel that feedback hub entries with API proposals are responded to, they seem to get lost in the stream of other feedback.

**Comment by hansmbakker** [comment]
This proposal was moved from https://github.com/microsoft/microsoft-ui-xaml/issues/2548 as requested by @ranjeshj

**Comment by shaheedmalik** [comment]
Good idea.

**Comment by hansmbakker** [comment]
Added some requirements

**Comment by jonaskuske** [comment]
Would love to see this! It should make it possible to write a plugin for the [Logitech Craft keyboard and its rotating crown](https://github.com/Logitech/logi_craft_sdk) that would enable support in all apps that don't use logitechs custom API/SDK, but *do* have support for Surface Dial right?

**Comment by hansmbakker** [comment]
Yes! That’s the idea.

There are more devices on the market that could act like a surface dial.

The benefit of translating the input to RadialController input is that it is standardized by Microsoft, so this way you can use your device without app developers having to use an app-specific SDK indeed.

Also, this way it is possible to build an emulator for Surface Dial in case you want to test your app for RadialController input.

**Comment by hansmbakker** [comment]
Developing a windows driver is another possibility (it can be a virtual hid driver in user mode) but it introduces lots of hurdles:
- Steep development learning curve for people coming from .Net development 
- less easy development loop (compile —> install —> install —> debug)
- Less easy debugging
- driver development tooling is not as user friendly as .net tooling

**Comment by feisu-ms** [comment]
Thanks for your interest – my team works on input devices like the Dial and we’re happy to see the interest here. The output vibration command part is unique for Dial compared to other input types and requires new OS support. We’re looking into a more generic input injection model for Windows, with device-specific helpers delivered here through Project Reunion.

In the meantime, if anyone has a particular Dial injection API shape in mind, please feel free to sketch it out.

**Comment by hansmbakker** [comment]
> The output vibration command part is unique for Dial compared to other input types
Understood. Is there a way that the input part could be made available before the output part is ready? That way, I believe a part of the people would be helped already.

> if anyone has a particular Dial injection API shape in mind

I suppose there of course needs to be an object describing input events, which could be like this:
```csharp
struct RadialControllerInputData
{
    bool ButtonPressed; //true is down / activated, false is up / untouched
    double (Delta)Rotation; //delta rotation in degrees, to make it device-independent. Should be signed to distinguish CW/CCW. 

    //RadialController has support for X and Y coordinates (where the controller should be shown on the screen), but personally I never used them. It could be added to be complete.
}
```

Following the existing [`InputInjector` API](https://docs.microsoft.com/en-us/uwp/api/windows.ui.input.preview.injection.inputinjector?view=winrt-19041) that would look something like this:
```csharp
public void InjectRadialControllerInput(IEnumerable<RadialControllerInputData> input);
```

**Comment by weltkante** [comment]
Maybe something like the UWP MIDI API? That one also supports different kinds of data packets, both for input and output.

**Comment by hansmbakker** [comment]
Added a requirement about mimicking the presence / disappearing of a RadialController.

This is needed because Windows shows / hides the `Wheel` page in the Settings app based on whether a Surface Dial is connected.

![image](https://user-images.githubusercontent.com/3463496/84531622-29620780-ace5-11ea-88b7-70d81b13494d.png)

**Comment by feisu** [comment]
> > The output vibration command part is unique for Dial compared to other input types
> 
> @feisu-ms Understood. Is there a way that the input part could be made available before the output part is ready? That way, I believe a part of the people would be helped already.
> 
> > if anyone has a particular Dial injection API shape in mind
> 
> I suppose there of course needs to be an object describing input events, which could be like this:
> 
> ```cs
> struct RadialControllerInputData
> {
>     bool ButtonPressed; //true is down / activated, false is up / untouched
>     double Rotation; //delta rotation in degrees, to make it device-independent. Should be signed to distinguish CW/CCW. 
> 
>     //RadialController has support for X and Y coordinates (where the controller should be shown on the screen), but personally I never used them. It could be added to be complete.
> }
> ```
> 
> Following the existing [`InputInjector` API](https://docs.microsoft.com/en-us/uwp/api/windows.ui.input.preview.injection.inputinjector?view=winrt-19041) that would look something like this:
> 
> ```cs
> public void InjectRadialControllerInput(IEnumerable<RadialControllerInputData> input);
> ```

That looks reasonable. We may add output support for haptics but agreed that's lower priority. Thanks.

**Comment by feisu** [comment]
> Maybe something like the UWP MIDI API? That one also supports different kinds of data packets, both for input and output.

Good point. Input is mostly HID based. But there are similarity and may be patterns we can leverage.

**Comment by feisu** [comment]
> Added a requirement about mimicking the presence / disappearing of a RadialController.
> 
> This is needed because Windows shows / hides the `Wheel` page (shown below) in the Settings app based on whether a Surface Dial is connected.
> 
> ![image](https://user-images.githubusercontent.com/3463496/84531622-29620780-ace5-11ea-88b7-70d81b13494d.png)

Injection devices have its own lifetime that's tied to the lifetime of the injecting app. If no app is injecting Dial, then there's no Dial on the system and thus no Settings for it.

**Comment by hansmbakker** [comment]
(Also discussed with @qmatteoq)

**Comment by hansmbakker** [comment]
@feisu 
> We may add output support for haptics but agreed that's lower priority

To give a bit more context: there are other devices than Surface Dial that have a form of output as well - mine can use a LED matrix to notify the user .

**Comment by hansmbakker** [comment]
@feisu could you give some indication on next steps / what we can expect?

**Comment by feisu** [comment]
@hansmbakker This is currently in our back log. It's unclear at this moment when we will be able to get to it. I will update this thread as soon as we have some update.

**Comment by hansmbakker** [comment]
Is there any update on this? I see the WinUI team published a roadmap update that mentions input (without specifying what that entails), does that mean anything for this proposal?

**Comment by stevewri** [comment]
@feisu - do you have any updates to share?

**Comment by feisu** [comment]
Nothing specific to share at the moment. This is still in our backlog. Compared to other input modalities like mouse/keyboard/touch/pen/touchpad, the RadialController device like Surface Dial is relatively less commonly used and thus we are prioritizing accordingly.

**Comment by hansmbakker** [comment]
@feisu although initially I understand that point, it is a bit of a chicken-egg problem.

Currently the only device that supports the RadialController APIs natively is the surface dial itself. Other similar devices exist, using other protocols.

If you would allow us to emulate such a device (by making existing devices compatible with the RadialController APIs by using a to-be-developed injection API), then more users could make use of apps that already support a surface dial, and then it could also be more interesting for developers to add support in their apps.

**Comment by feisu** [comment]
@hansmbakker you made a very good point and it's very much appreciated. We will definitely take your comments into account in our prioritization.
